cpp/include/cudf/detail/gather.cuh

```diff
@@ -255,7 +259,455 @@ struct column_gatherer_impl<dictionary32, MapItType> {
+
+    // offsets of the source column
+    size_type const* src_offsets{source_column.offsets().data<size_type>()};
+
+    // outgoing offsets.  these will persist as output from the entire gather operation
+    auto dst_offsets_c = cudf::make_fixed_width_column(
+      data_type{INT32}, offset_count, mask_state::UNALLOCATED, stream, mr);
+    mutable_column_view dst_offsets_v = dst_offsets_c->mutable_view();
+    size_type* dst_offsets            = dst_offsets_v.data<size_type>();
+
+    // generate the compacted outgoing offsets.
+    auto count_iter = thrust::make_counting_iterator<size_type>(0);
+    thrust::plus<size_type> sum;
+    thrust::transform_exclusive_scan(
+      rmm::exec_policy(stream)->on(stream),
+      count_iter,
+      count_iter + offset_count,
+      dst_offsets,
+      [gather_map, output_count, src_offsets] __device__(size_type index) -> size_type {
+        // last offset index is always the previous offset_index + 1, since each entry in the gather
+        // map represents a virtual pair of offsets
+        size_type offset_index =
+          index < output_count ? gather_map[index] : gather_map[index - 1] + 1;
+        // the length of this list
+        return src_offsets[offset_index + 1] - src_offsets[offset_index];
+      },
+      0,
+      sum);
+
+    // for each span of offsets (each output row) we need to know the original offset value to build
+    // the gather map for the next level.  this data is temporary and will only persist until the
+    // next level of recursion is done using it.  This way, even if we have a hierarchy 1000 levels
+    // deep, we will only ever have at most 2 of these extra columns in memory.
+    auto base_offsets = cudf::make_fixed_width_column(
+      data_type{INT32}, output_count, mask_state::UNALLOCATED, stream, mr);
+    mutable_column_view base_offsets_mcv = base_offsets->mutable_view();
+    size_type* base_offsets_p            = base_offsets_mcv.data<size_type>();
+
+    // generate the base offsets
+    thrust::transform(rmm::exec_policy(stream)->on(stream),
+                      gather_map,
+                      gather_map + offset_count,
+                      base_offsets_p,
+                      [src_offsets] __device__(size_type index) { return src_offsets[index]; });
+
+    return {std::move(dst_offsets_c), std::move(base_offsets)};
+  }
+
+  /**
+   * @brief The information needed to create an iterator to gather level N+1
+   *
+   * See documentation for make_gather_offsets for a detailed explanation.
+   */
+  struct gather_data {
+    size_type const* offsets;
+    std::unique_ptr<column> base_offsets;
+  };
+
+/**
+ * @brief Macro for creating a gather map iterator to gather from level N+1.
+ *
+ * The iterator for each level is unique and we can't use templates because the
+ * operation is recursive.  There is a restriction in nvcc that prevents you
+ * from returning a __device__ lambda from a function, so this cannot be implemented
```

Comment:
I'm pretty strongly opposed to using a macro like this. This is going to require a lot more exploration and justification before moving forward with it.

Reply:
I haven't looked at this in detail yet, but if you can't use a lambda, what about a function object?

Reply:
I'll take a look.  I figured this was pretty hokey.  

As a side note, some of the code has been rearranged out of detail/gather.cuh and into  lists/detail/gather.cuh and lists/copying/gather.cu.   It didn't need to be in such a big blob in this one place (particularly the non-template functions).